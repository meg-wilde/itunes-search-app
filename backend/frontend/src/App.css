/* header styles */
h1,
h2 {
  text-align: center;
  font-family: Arial, Helvetica, sans-serif;
  color: #03045e;
}

/* Style for layout of search box section */
#search-boxes {
  display: flex;
  justify-content: center;
  align-items: center;
}

/* style for the actual search section components */
input,
select,
button {
  border-radius: 5px;
  border-width: 1px;
  border-color: #03045e;
  padding: 8px;
  margin: 10px;
  margin-top: 20px;
}

/* button styling including hover state, clicked (active) and deactivates state */
button {
  background-color: #00b4d8;
  color: #caf0f8;
}

button:hover {
  color: #03045e;
}

button:active {
  background-color: #0077b6;
  color: #03045e;
}

button:disabled {
  background-color: grey;
  color: #caf0f8;
  cursor: not-allowed;
}

/* create layout for the results and favourites sections */
.results-container,
.favourites-container {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  gap: 20px;
  margin: 20px;
}

/* style for each result box */
.result-item,
.favourite-item {
  border: 1px solid #ddd;
  padding: 10px;
  text-align: center;
}

/* style for image for each result */
.result-item img,
.favourite-item img {
  width: 200px;
  height: 200px;
  object-fit: cover;
}

/* Nav bar background color, height and font alignment*/
nav {
  background-color: #90e0ef;
  font-family: Arial, Helvetica, sans-serif;
  font-size: 18px;
  text-align: center;
  height: 40px;
  width: 70%;
  margin: 0 auto;
  border-radius: 10px;
}

/* Nav bar elements positioning*/
nav li {
  margin-top: 8px;
  margin-right: 10px;
  display: inline-block;
}

/* Nav bar text colour*/
.link {
  color: #03045e;
  text-decoration: none;
}

/* Change colour on hover*/
.link:hover {
  color: white;
  text-decoration: none;
}
