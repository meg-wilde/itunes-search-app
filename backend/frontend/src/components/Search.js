import React, { useState } from "react";

//component to search for the item
const Search = ({ onSearch }) => {
  const [term, setTerm] = useState(""); //set the term to blank
  const [mediaType, setMediaType] = useState(""); //set the media to blank

  //function to handle the search
  const handleSearch = () => {
    //take in the term and media type
    onSearch(term, mediaType);
  };

  return (
    <div id="search-boxes">
      {/* create a text input box to take the search term */}
      <input
        type="text"
        placeholder="Enter search term"
        value={term}
        onChange={(e) => setTerm(e.target.value)} //when its filled in set the term to the value
      />
      {/* create a select dropdown for the mediatype and set it to the value selected */}
      <select value={mediaType} onChange={(e) => setMediaType(e.target.value)}>
        {/* list of options for mediaType */}
        <option value="music">Music</option>
        <option value="movie">Movie</option>
        <option value="podcast">Podcast</option>
        <option value="ebook">Ebook</option>
        <option value="musicVideo">musicVideo</option>
        <option value="audiobook">Audio Book</option>
        <option value="shortFilm">Short Film</option>
        <option value="tvShow">TV Show</option>
        <option value="software">Software</option>
        <option value="all">all</option>
      </select>
      {/* when the search button is clicked trigger the handleSearch function */}
      <button onClick={handleSearch}>Search</button>
    </div>
  );
};

export default Search;
