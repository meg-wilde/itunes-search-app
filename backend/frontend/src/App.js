import React, { useState, useEffect } from "react";
import Search from "./components/Search";
import Results from "./components/Results";
import Nav from "./components/Nav";
import Favourites from "./components/Favourites";
import "./App.css";

function App() {
  const [searchResults, setSearchResults] = useState([]); //create a blank array as the initial state for the search results
  const [favourites, setFavourites] = useState([]); //create a blank array to start for storing favourites

  const handleSearch = async (term, mediaType) => {
    // Make the API request using the url set above
    //take the search term and the media type and pass them into the search
    const url = `/search?term=${term}&media=${mediaType}`;
    const response = await fetch(url); //wait for the response
    const data = await response.json(); //save the response json as data
    setSearchResults(data.results || []); //add the data results into the array for search results
  };

  //function to add a result to favourites
  const addToFavourites = (result) => {
    const { trackId, trackName, artistName, artworkUrl100 } = result; //add the Id, name, artist and image to the result

    let myFavourites = JSON.parse(localStorage.getItem("myFavourites")) || []; //save the favourites to local storage

    // Check if the item has already been saved by checking if the id, name artist and image all match
    const isAlreadySaved = myFavourites.some(
      (item) =>
        item.trackId === trackId &&
        item.trackName === trackName &&
        item.artistName === artistName &&
        item.artworkUrl100 === artworkUrl100
    );

    if (!isAlreadySaved) {
      // If the item has not already been saved
      const newFavourite = { trackId, trackName, artistName, artworkUrl100 }; //create a new item
      //add the newFavourite to myFavourites
      myFavourites = [...myFavourites, newFavourite];
      localStorage.setItem("myFavourites", JSON.stringify(myFavourites));
      setFavourites(myFavourites); // Update state to trigger re-render
    } else {
      // If the item has already been saved, do nothing
    }
  };

  //function to remove item from favourites
  const removeFromFavourites = (result) => {
    const { trackId, trackName, artistName, artworkUrl100 } = result;
    // Update state to filter out and remove the item from the displayed favourites
    setFavourites((prevFavourites) =>
      prevFavourites.filter(
        (favourite) =>
          !(
            favourite.trackId === trackId &&
            favourite.trackName === trackName &&
            favourite.artistName === artistName &&
            favourite.artworkUrl100 === artworkUrl100
          )
      )
    );

    // Update localStorage after removal
    const updatedFavourites =
      JSON.parse(localStorage.getItem("myFavourites")) || [];
    // Filter out the removed item from the stored favourites
    const newFavourites = updatedFavourites.filter(
      (favourite) =>
        !(
          favourite.trackId === trackId &&
          favourite.trackName === trackName &&
          favourite.artistName === artistName &&
          favourite.artworkUrl100 === artworkUrl100
        )
    );
    // Save the updated favourites back to localStorage
    localStorage.setItem("myFavourites", JSON.stringify(newFavourites));
  };

  useEffect(() => {
    // Define a function to fetch favourites from local
    const fetchFavourites = () => {
      // Retrieve the stored favourites from local
      const storedFavourites = JSON.parse(localStorage.getItem("myFavourites"));
      // Set the favourites state to the retrieved favourites or an empty array if none exists
      setFavourites(storedFavourites || []);
    };
    // Call the fetchFavourites function when the component mounts
    fetchFavourites();
  }, []);

  //page layout
  return (
    <div>
      {/* always show the H1 and the nav */}
      <h1>iTunes Search API</h1>
      <Nav />
      {window.location.pathname === "/" ? (
        <div>
          {/* if the url path is "/" then show the search and results components */}
          <Search onSearch={handleSearch} />
          <Results
            results={searchResults}
            addToFavourites={addToFavourites}
            favourites={favourites}
          />
        </div>
      ) : window.location.pathname === "/favourites" ? (
        <div>
          {/* if the url is /favourites when how the favourites component */}
          <Favourites
            favourites={favourites}
            removeFromFavourites={removeFromFavourites}
          />
        </div>
      ) : (
        // if another path then show page not found
        <p>Page not found</p>
      )}
    </div>
  );
}

export default App;
