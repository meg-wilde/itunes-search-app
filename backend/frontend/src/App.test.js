import React from "react";
import { render } from "@testing-library/react";
import Search from "../src/components/Search";
import App from "./App";

//create a snapshot test for the search component
test("renders correctly", () => {
  const { asFragment } = render(<Search />);
  expect(asFragment()).toMatchSnapshot();
});

//unit test to test the whole app and see if it renders.
test("renders App component without crashing", () => {
  render(<App />);
});
