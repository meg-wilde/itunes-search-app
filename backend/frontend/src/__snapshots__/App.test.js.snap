// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    id="search-boxes"
  >
    <input
      placeholder="Enter search term"
      type="text"
      value=""
    />
    <select>
      <option
        value="music"
      >
        Music
      </option>
      <option
        value="movie"
      >
        Movie
      </option>
      <option
        value="podcast"
      >
        Podcast
      </option>
      <option
        value="ebook"
      >
        Ebook
      </option>
      <option
        value="musicVideo"
      >
        musicVideo
      </option>
      <option
        value="audiobook"
      >
        Audio Book
      </option>
      <option
        value="shortFilm"
      >
        Short Film
      </option>
      <option
        value="tvShow"
      >
        TV Show
      </option>
      <option
        value="software"
      >
        Software
      </option>
      <option
        value="all"
      >
        all
      </option>
    </select>
    <button>
      Search
    </button>
  </div>
</DocumentFragment>
`;
